src/TMxpCustomElementTagHandler.cpp

```diff
@@ -89,11 +89,22 @@ MxpStartTag TMxpCustomElementTagHandler::resolveElementDefinition(const TMxpElem
         if (!attr.hasValue()) {
             return MxpTagAttribute(mapAttributes(element, attr.getName(), customTag));
         } else {
+
+/*
+ * Michael Weller: What is this? Why are hints forced to be all upper, even when
+ * explicitly set by the Mud in mixed case?
+ * And only when <send> is used in a custom element, but not when used directly?
+ * Let the mud choose about its user interface.
+ */
+#if 0
```

Comment:
Are you sure it's forced and hasn't been converted to uppercase elsewhere?

Reply:
Yes, 100%:
When using mudlet to connect to our mud I was surprised / wondering why all MXP menus appear in uppercase, though they are specified with mixed case in the mud and display as such in all other mud clients.
The patch above definitely changes this to behave 'normal'.
IMHO, it's fine when a mud wants to display MXP hints in all upper as a 'design' decision, it can just specify the hints in uppercase. But - I think the mud client should not enforce an all upper display on its own for stylistic reasons - Or if this is really something players want to have, it should be an option in the game profile - like a tick box: 'display MXP send menus in all uppercase'.
I was even more surprised, that the uppercase conversion only occurs for send-hints when send is used as a part of a custom element tag. It does not occur when send is used directly. I didn't notice this fine distinction at first because we use custom elements all over. But when I saw it in the source, I could prove it easily with an mxp testtool.

Reply:
Yeah I agree with you, and I think we haven't noticed it because it happens in a specific circumstances as you say. Not intentional. I think you can take the comment out

Reply:
Ok, removed the comment, github tests ran fine.

Reply:
Excellent. Speaking of the tests, can you add ones to cover your improvements? MXP is hairy and adjusting it for one game can easily break another, so we're starting to lock it down with tests.

Reply:
Yes, definitely. But I'm not sure what exactly you have in mind. Ideally, there should be some test file(s) with
different MXP scenarios (text files, basically) and then one should check if mudlet does not crash or lockup when
rendering them and more difficult, that the result looks right. Even more difficult, some things depend on a user
action (hover over a link, select a menu). Imho that's difficult to automate, not impossible (maybe for me w/o a proper tool). Is there something in place for mudlet already, which I'm not aware off?

Probably it'll be easier if one could load a file into mudlet and see how it displays and compare with an example document (which might include screen shots; - even from other clients). I assume you might be able to do this with a lua script, but it won't test the complete input chain. It would be better to connect to a mudstyle server thatjust produces the test pattern. There is this mudlet test server. I assume it does something like this, but I can't connect to. I checked my IPv6 connection with other sites, it worked, but I can't reach the test server.

I'd be willing to provide some test patterns/files, surely - I have one for my own experiments already and can enhance them. Even for things that don't (yet) work with mudlet, even lock it up (esp when a mud misbehaves). But I'd need to know exactly what you need, a text file, a doc file with example screen shots, etc..

If there's a hint in the mudlet docs or web site regarding this, I didn't find it, so just give a pointer.

Michael.

Reply:
That sounds great! There's a few ways of doing them, we've went along the C++/API level. You can find the tests in https://github.com/Mudlet/Mudlet/tree/development/test and you can see them running for your PR in [here](https://github.com/Mudlet/Mudlet/pull/4200/checks?check_run_id=1329194065):

![image](https://user-images.githubusercontent.com/110988/97748370-49a03080-1ae5-11eb-926d-459aeb18bed4.png)

Reply:
Sorry for the late reply, I was busy in RL. I need a helping hand, as I don't get QT Creator (on Win10) to even try compiling the test directory. What I found in the manuals didn't work out, though I tried a bunch of things, but it doesn't want to compile/perform any tests.

Regarding the tests themselves.. I'm well, uncertain on my next steps. I wrote already I consider making sensible tests really difficult. 

These tests.. I'm really speechless... well they are bare unit tests, they can test some internal functions and that's it. I fear they are actually counterproductive as they might make people think MXP is working and well tested. I was quite shocked when I looked into the Mudlet src and see that most of the MXP functionality missing from the released program is already there, or at least already laid out in the design. However, the functions are not called right by the code around them or don't have any effect beyond parsing their arguments. I can now clearly see how this situation developed. 

What I don't see is how a unit level test will help to check for or mitigate these problems. They'd help if I made up some complex internal memory structure and want some automated test calls to run and check if my memory handling is right. But checking for proper interaction with TTextedit?

I need to be able to compile the test directory, then I can see if I can make up something. However, I fear whatever I do will hinder someone else from improving my code and mudlet in general, when he - for example - optimizes some internals and made them work differently. This is explicictly not what I want to do.

Reply:
Yeah, happy to lend a hand in getting it to run the tests. Jump in on Discord.

Mudlet by design implements a small subset, not a full version, of MXP. Very few games use the full blown MXP implementations and since the standard is a good mess, we don't deal with it besides the bare minimum. That's why MXP was rewritten, only a small amount - what we actually supported before - was actually plugged in, and the rest was left available, which is what you came across.

Reply:
Is it really required to use Discord? It can't be any more than a few check marks or options to make, which I just can't figure out w/o reading the whole lot of Qt docs and not wanting to risk breaking the projects definition files. I'm not into socialising by PC and I don't feel like getting a mic/headset and install discord just for a simple bugfix.

Reply:
Oh. We actually use discord mostly for quick text chat much like irc but with inline screenshots, etc.  However hardly use the voice chat you assume. Also it runs in a browser, no need to install anything.

Reply:
Yeah it's our replacement nforbchat really. That said I'll look into documenting how to run the tests as well.

Reply:
Yeah, for the record we used [Gitter](https://gitter.im/Mudlet/Mudlet) a couple of years back but now the only activity you'll see there is notifications from the CI builds - otherwise it is all tumbleweed there now.

You can use this permanent [invite](https://discord.com/invite/kuYvMQ9) from Vadim to access our Discord guild...

Reply:
>  That said I'll look into documenting how to run the tests as well.

Yep, that would be most helpful and highly appreciated. The docs on how to rebuild mudlet were really helpful, btw. There was one or the other gap to my level of Github and QT knowledge, but it could be bridged. From what I read and understood, I just need a different project / mudlet.po file, or clone the existing and add one or more build parameters which I'm unable to figure out. I already enabled "running all tests with every build" in QTCreator, but it just does not seem to have any effect.

Reply:
I think you need a cmake build for the tests and the instructions currently use qmake. Are you on Windows?

Reply:
Yes, I'm on windows (If anyone had told me I build something under Windows.. I'm really a Unix/Linux guy at least I was in the ancient days when I had time, but Windows is what I use for daily work now) . And yes, qmake or actually: C:\Qt\5.13.2\mingw73_32\bin\qmake.exe and/or C:\Qt\Tools\mingw730_32\bin\mingw32-make.exe are used. I can change the latter make executable to C:\Program Files (x86)\CMake\bin\cmake.exe in the project, and cmake is used. However, this cmake does not accept the -j<proc> parameter. I can change the <proc> parameter, but I cannot remove it.
Also, trying to do things on cli, it seems cmake want to make a build, not an executable. So probably I need to replace the qmake be cmake, but I cannot change that in the project settings.
The same directory has also a ctest, but it complains there is no test configuration file found. QT Creator, BTW, has a bunch of options running tests, but it seems, no test config is found. Hence, QT Creator does not run any tests.

Reply:
Well, I managed to get it to compile with qmake and mingw32-make as installed according to your 'Compiling on Windows 7+' instructions. Unfortunately I had to learn that qmake does not really support building multiple targets. So I had to play around with creating a bunch of makefiles on the fly, etc. However with a project file (disguised with an uncommon name s.t. QtCreator), a plain Makefile to make the makefiles and a batch file to set the paths as needed (the Qt installation didn't extend the system paths.. maybe a good thing (maybe not)). Basically you 'RunTestsOnWindows' in the test directory and thats it. 

Hmm.. I can't attach the files to this post, so I'll include them to the pull request, for you to have a look.

You should think about and give me an indication, if you want my enhanced MXP tests as part of this pull request, or yet another one (IMHO, another one makes little sense, as my tests, when sensible, will definitely fail w/o my patches)

I will now:

a) Fix the commented out TMxpSendTagHandlerTest.cpp s.t. it works again.

b) Try to add some more, sensible tests. I have some ideas to make them at least somewhat sensible. However, as I already tried to elaborate: There was (almost) no issue with parsing MXP, hence I didn't change more than a few tiny bits of the parser, the issues were in how MXP functions were called from the rest of the code (or weren't) and how they interacted with TBuffer and friends. IMHO, any at least barely sensible test would include some kind of 'fake' TBuffer which will record the outputs of the MXP stages send to the display in some kind of text form which can then checked for. IMHO, a sensible test of MXP (or any other aspect of the terminal emulation) would include some 'mud' input, ideally coming in through all layer of the tcp and telnet handling and then checking the visual output of the mudlet window... where actually, OS and version specific variations regarding fonts, exact placement of tooltips.. exact 'RGB' values of standard colours... and and and are OS and implementation specific, hence slight variations are totally fine.

Well...
Enjoy.

Reply:
Hi Vadim,
We have some serious OS/ISP Problems at our game that will require my undivided attention / free time for several days, maybe weeks. Due to that I cannot currently enhance the test code. I'll get back checking for this, but I cannot yet foresee when.